Cleanup MobileCore APIs

praveek opened the pull request:
Cleanup MobileCore APIs to use the new Eventhub. 

#### New API

Added new API so that extensions can explicitly specify timeout for each call. 
`public static void dispatchEventWithResponseCallback(final Event event, long timeoutMS, final AdobeCallbackWithError<Event> responseCallback)`

#### Deprecated APIs

These APIs were only used by Extensions. Going forward, they can use equivalent APIs in ExtensionAPI class. 
`public static void dispatchEventWithResponseCallback(final Event event,
                                                         final AdobeCallbackWithError<Event> responseCallback)`

`public static boolean dispatchEvent(final Event event, final ExtensionErrorCallback<ExtensionError> errorCallback)`

`public static boolean dispatchEventWithResponseCallback(final Event event,
                                                            final AdobeCallback<Event> responseCallback,
                                                            final ExtensionErrorCallback<ExtensionError> errorCallback)`

`public static boolean dispatchResponseEvent(final Event responseEvent, final Event requestEvent,
                                                final ExtensionErrorCallback<ExtensionError> errorCallback)`

Added seperate issues to fix
1) [Cleanup App class used to store ApplicationContext](https://github.com/adobe/aepsdk-core-android/issues/151)
2) [Remove MobileCore.getEventHistory API. ](https://github.com/adobe/aepsdk-core-android/issues/150)

## Description

<!--- Describe your changes in detail -->

## Related Issue

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Screenshots (if appropriate):

## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

yangyansong-adbe commented on code/android-core-library/src/main/java/com/adobe/marketing/mobile/internal/eventhub/EventHubExtensionRegistarar.kt:
```diff
@@ -0,0 +1,90 @@
+package com.adobe.marketing.mobile.internal.eventhub
```
Missing copyright and license headers.

yangyansong-adbe commented on code/android-core-library/src/main/java/com/adobe/marketing/mobile/internal/eventhub/EventHubExtensionRegistarar.kt:
```diff
@@ -0,0 +1,90 @@
+        synchronized(mutex) {
+            if (!startCalled) {
+                extensionsRegisteredBeforeStart.add(extensionClass)
+            }
```
Should we return and log a warning message if SDK was started?

yangyansong-adbe commented:
IMO, the `pendingHubStart` class is way too complicated. Can we just use an atomic boolean and a concurrent list to guard the `Eventhub start` and the `register extension` operations? If the Eventhub was started the later start/registerExtension calls should be ignored. If the extension registration failed,  we don't even need to remove it from the registered extension list as we don't support registering the same extension multiple times.

praveek commented:
I agree. I added the logic to support current registration logic. I have removed `EventHubExtensionRegistarar` class for now and will revisit as part of https://github.com/adobe/aepsdk-core-android/issues/155

yangyansong-adbe commented on code/android-core-library/src/main/java/com/adobe/marketing/mobile/Extension.java:
```diff
@@ -58,7 +58,7 @@ protected String getVersion() {
 	 * Optional metadata provided for use by the event hub for logging.
 	 *
-	 * @return the extension metadata as a {@link Map<String, String}
+	 * @return the extension metadata as a {@code Map<String, Object>}
```
The `getMetadata` method returns an instance of `Map<String, String>`.

spoorthipujariadobe replied:
Return type is Map<String, String>, typo in javadocs

yangyansong-adbe commented on code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:
```diff
@@ -250,118 +193,103 @@ public static void configureWithAppID(final String appId) {
-            Log.debug(LOG_TAG, "Failed to dispatch event. (%s)", NULL_CONTEXT_MESSAGE);
+    public static void start(final AdobeCallback<?> completionCallback) {
+        if(!sdkInitializedWithContext.get()) {
+            Log.error(LOG_TAG, "Failed to start - completionCallback is null");
```
We need to update this error log.

yangyansong-adbe commented on code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:
```diff
@@ -373,153 +301,227 @@ public void call(final Event e) {
+     * @param appId A unique identifier assigned to the app instance by Adobe Launch
+     */
+    public static void configureWithAppID(final String appId) {
+        Map<String, Object> eventData = new HashMap<>();
```
Can we add a null/empty check here?

yangyansong-adbe commented on code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:
```diff
@@ -373,153 +301,227 @@ public void call(final Event e) {
+     * @param fileName the name of the configure file in the assets folder. A value of {@code null} has no effect.
+     */
+    public static void configureWithFileInAssets(final String fileName) {
+        Map<String, Object> eventData = new HashMap<>();
```
null/empty check?

yangyansong-adbe commented on code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:
```diff
@@ -529,21 +531,13 @@ public static void configureWithFileInAssets(final String fileName) {
-            Log.warning("Configuration", "Unable to configure with null or empty file path");
-            return;
-        }
+        Map<String, Object> eventData = new HashMap<>();
```
null/empty check?

yangyansong-adbe commented on code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:
```diff
@@ -559,19 +553,13 @@ public static void configureWithFileInPath(final String filepath) {
-        final Event event = new Event.Builder("Configuration Update", EventType.CONFIGURATION,
-                EventSource.REQUEST_CONTENT).setData(eventData).build();
-        eventHub.dispatch(event);
+        Map<String, Object> eventData = new HashMap<>();
```
null/empty check?

yangyansong-adbe commented on code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:
```diff
@@ -846,195 +729,201 @@ public static void lifecycleStart(final Map<String, String> additionalContextDat
-        eventHub.dispatch(new Event.Builder("CollectPII", EventType.GENERIC_PII,
-                EventSource.REQUEST_CONTENT).setData(eventData).build());
-        Log.trace(LOG_TAG, "Collect PII event was sent");
+    public static void trackAction(final String action, final Map<String, String> contextData) {
```
null/empty check?

praveek replied:
I will check with the Analytics squad and update trackState/trackAction later

yangyansong-adbe commented on code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:
```diff
@@ -846,195 +729,201 @@ public static void lifecycleStart(final Map<String, String> additionalContextDat
-    public static void setSmallIconResourceID(final int resourceID) {
-        App.setSmallIconResourceID(resourceID);
+    public static void trackState(final String state, final Map<String, String> contextData) {
+        Map<String, Object> eventData = new HashMap<>();
```
null/empty check?

yangyansong-adbe reviewed (approved):
LGTM, only several small comments.

yangyansong-adbe commented:
[Improve code inspection with annotations 
](https://developer.android.com/studio/write/annotations)
@praveek one more question, can we consider adding `@Nullable`/`@NonNull` annotations to the public APIs in SDK 2.0?

spoorthipujariadobe commented on code/android-core-library/src/main/java/com/adobe/marketing/mobile/EventType.java:
```diff
@@ -27,6 +27,12 @@ public final class EventType {
 	public static final String TYPE_WILDCARD = "com.adobe.eventType._wildcard_";
 	public static final String TYPE_HUB = "com.adobe.eventType.hub";
+	public static final String TYPE_CONFIGURATION = "com.adobe.eventType.configuration";
+	public static final String TYPE_GENERIC_LIFECYLE = "com.adobe.eventType.generic.lifecycle";
```
Typo here. Should be `TYPE_GENERIC_LIFECYCLE`

spoorthipujariadobe commented on code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:
```diff
@@ -250,118 +193,103 @@ public static void configureWithAppID(final String appId) {
 
     /**
      * This method will be used when the provided {@code Event} is used as a trigger and a response event
      * is expected in return. The returned event needs to be sent using
```
Need to update javadocs here. `dispatchEvent` does not expect a response event in return right?

spoorthipujariadobe commented on code/android-core-library/src/phone/java/com/adobe/marketing/mobile/MobileCore.java:
```diff
@@ -580,17 +568,13 @@ public static void updateConfiguration(final Map<String, Object> configMap) {
-                EventSource.REQUEST_CONTENT).setData(eventData).build();
-        eventHub.dispatch(event);
+
+        Event event = new Event.Builder("Configuration Update", EventType.CONFIGURATION,
```
Event name is changed, need to update `MobileCoreTests` if we want to use the new event name

spoorthipujariadobe replied:
Nvm, just saw you have changed the tests
